aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -1601,6 +1601,154 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
   return linalg_norm_out_impl(result, self, c10::nullopt, ord, opt_dim, keepdim, opt_dtype);
 }
 
+Tensor _linalg_cond_exception_helper(const Tensor& self) {
+  // For batched input if at least one matrix in the batch is not invertible,
+  // we can't get the result for all other (possibly) invertible matrices in the batch without an explicit for loop.
+  // This should change when at::inverse works with silent errors
+  if (self.dim() > 2) {
+    TORCH_CHECK(false,
+      "At least one matrix in the batch is not invertible, its condition number is infinity, "
+      "linalg_cond does not support yet calculating the condition number for all other (possibly invertible) matrices in the batch.");
+  }
+  auto result_shape = self.sizes().vec();
+  result_shape.pop_back();
+  result_shape.pop_back();  // result's shape is equal to self.shape[0:-2]
+  Tensor result = at::empty(result_shape, self.options());
+  at::fill_(result, INFINITY);
+  return result;
+}
+
+// Numerical or None norms
+Tensor linalg_cond(const Tensor& self, optional<Scalar> opt_ord) {
+  TORCH_CHECK(self.numel() > 0, "linalg_cond is not defined for empty tensors.");
+  TORCH_CHECK(self.dim() >= 2, "linalg_cond only supports matrices or batches of matrices, but got a tensor with ",
+    self.dim(), " dimensions.");
+
+  // The default case is using 2-norm
+  Scalar ord = opt_ord.has_value() ? opt_ord.value() : 2;
+
+  // If ord == None or ord == ±2
+  if (std::abs(ord.toDouble()) == 2.0) {
+    auto singular_values = std::get<1>(at::svd(self));
+    auto s_max = std::get<0>(singular_values.max(/*dim=*/-1));
+    auto s_min = std::get<0>(singular_values.min(/*dim=*/-1));
+    Tensor result;
+    if (ord.toDouble() == -2.0) {
+      result = s_min / s_max;
+    } else {
+      result = s_max / s_min;
+    }
+    // convert FLT_MAX or DBL_MAX to INFINITY for NumPy compatibility
```

Comment:
Should we let ourselves be divergent from NumPy here?

This is going to be a significant perf hit. And what if the condition number actually is double max? 

Can you elaborate on what kind of issues we'd run into if we didn't do this conversion?

Reply:
When the matrix is non-invertible (singular) it means it has zero singular values. Singular values are always non-negative, so the minimal possible value is zero. Now when we divide something with 0 we should get inf (by IEEE 754), but at the C++ level this gives FLT_MAX or DBL_MAX.
While in Python everything seems to be as expected.
```python
In [1]: import torch
In [2]: a = torch.eye(3)
In [3]: a[-1, -1] = 0  # make 'a' singular
In [4]: u, s, v = a.svd()
In [5]: s
Out[5]: tensor([1., 1., 0.])
In [6]: s.max() / s.min()
Out[6]: tensor(inf)
```

Reply:
Wait... you're saying that our division operator isn't IEEE 754 compliant?

Here's our CPU division kernel as a reference: https://github.com/pytorch/pytorch/blob/df0ae244a999a365ac13695cf4682ea30490b157/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp#L95

And our CUDA division kernel:
https://github.com/pytorch/pytorch/blob/df0ae244a999a365ac13695cf4682ea30490b157/aten/src/ATen/native/cuda/BinaryMulDivKernel.cu#L47

Can you elaborate on the behavior you're seeing on each device type?

Reply:
Alright, I'm sorry for the confusion. All this FLT_MAX/DBL_MAX conversion to inf is redundant here. I will remove it now. Of course the division operator is compliant with the IEEE 754.

For history:
Initially, I had only one return statement in the end and I had to take care of NANs from possibly non-invertible input. I decided to use `at::nan_to_num` assuming it does what's in the name - converting NANs to the specified value, BUT additionally, this function also converts +inf to FLT_MAX/DBL_MAX, I should have read the documentation. And then I "fixed" it by `result=result+result-result` for +inf this of course gave nan that is then converted back to +inf 😅. I mistakenly thought that the problem is somewhere else that lead to the current code with these redundant calls to `at::where(result == dbl_max, inf, result)`.

Reply:
Whew. Glad to hear it.

`nan_to_num`s behavior is confusing! Unfortunately our `nan_to_num` is consistent with NumPy's https://numpy.org/doc/stable/reference/generated/numpy.nan_to_num.html?highlight=nan_to_num#numpy.nan_to_num.